src/webgpu/shader/execution/shader_io/compute_builtins.spec.ts

```diff
@@ -254,3 +255,505 @@ g.test('inputs')
+  }
+  if (max < subgroupSize) {
+    return new Error(`Subgroup size '${subgroupSize}' is greater than maximum '${max}'`);
+  }
+
+  // Check that remaining invocations record a consistent subgroup size.
+  for (let i = 1; i < data.length; i++) {
+    if (data[i] !== subgroupSize) {
+      return new Error(
+        ErrorMsg(`Invocation ${i}: subgroup size inconsistency`, data[i], subgroupSize)
+      );
+    }
+  }
+
+  // Assumes workgroups are divided such that there is a single partial subgroup if any.
+  const partialExpected = invocations % subgroupSize;
+  let fullExpected = 0;
+  if (subgroupSize <= invocations) {
+    fullExpected = Math.floor(invocations / subgroupSize) * subgroupSize;
+  }
+  let fullCount = 0;
+  let partialCount = 0;
+  for (let i = 0; i < compare.length; i++) {
+    if (i % invocations === 0) {
+      if (i !== 0) {
+        // Check intermediate workgroup counts.
+        if (fullCount !== fullExpected) {
+          return new Error(
+            ErrorMsg(`Unexpected number of full invocations`, fullCount, fullExpected)
+          );
+        }
+        if (partialCount !== partialExpected) {
+          return new Error(
+            ErrorMsg(`Unexpected number of partial invocations`, partialCount, partialExpected)
+          );
+        }
+      }
+
+      // Reset for new workgroup.
+      fullCount = 0;
+      partialCount = 0;
+    }
+
+    if (compare[i] === subgroupSize) {
+      fullCount++;
+    } else {
+      partialCount++;
+
+      // Check that all partial invocations are consistent.
+      if (compare[i] !== partialExpected) {
+        return new Error(ErrorMsg(`Partial subgroup size incorrect`, compare[i], partialExpected));
+      }
+    }
+  }
+
+  // Check final workgroup counts.
+  if (fullCount !== fullExpected) {
+    return new Error(ErrorMsg(`Unexpected number of full invocations`, fullCount, fullExpected));
+  }
+  if (partialCount !== partialExpected) {
+    return new Error(
+      ErrorMsg(`Unexpected number of partial invocations`, partialCount, partialExpected)
+    );
+  }
+
+  return undefined;
+}
+
+const kWGSizes = [
+  [1, 1, 1],
+  [4, 1, 1],
+  [8, 1, 1],
+  [16, 1, 1],
+  [32, 1, 1],
+  [64, 1, 1],
+  [128, 1, 1],
+  [256, 1, 1],
+  [1, 4, 1],
+  [1, 8, 1],
+  [1, 16, 1],
+  [1, 32, 1],
+  [1, 64, 1],
+  [1, 128, 1],
+  [1, 256, 1],
+  [1, 1, 4],
+  [1, 1, 8],
+  [1, 1, 16],
+  [1, 1, 32],
+  [1, 1, 64],
+  [3, 3, 3],
+  [4, 4, 4],
+  [16, 16, 1],
+  [16, 1, 16],
+  [1, 16, 16],
+  [15, 3, 3],
+  [3, 15, 3],
+  [3, 3, 15],
+  [17, 5, 2],
+  [17, 4, 2],
+  [15, 2, 8],
+] as const;
+
+g.test('subgroup_size')
+  .desc('Tests subgroup_size values')
+  .params(u =>
+    u
+      .combine('sizes', kWGSizes)
+      .beginSubcases()
+      .combine('numWGs', [1, 2] as const)
+  )
+  .beforeAllSubcases(t => {
+    t.selectDeviceOrSkipTestCase('subgroups' as GPUFeatureName);
+  })
+  .fn(async t => {
+    // Replace these with the limits when they are wired up.
+    const minSize = 4;
```

Comment:
I thought at least in Chrome currently `device.limits.minSubgroupSize` is available as long as `subgroups` features is available?

Reply:
They are in Chrome, but not in dawn-node I believe which is significantly easier to for me to test.